modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/SleepTimer.kt

```diff
@@ -31,171 +30,191 @@ class SleepTimer @Inject constructor(
-    ) = diffTime < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && !lastEpisodeUuidAutomaticEnded.isNullOrEmpty() && currentEpisodeUuid != lastEpisodeUuidAutomaticEnded && !isSleepEndOfEpisodeRunning
-
-    private fun shouldRestartSleepEndOfChapter(diffTime: Duration, isSleepEndOfChapterRunning: Boolean) = diffTime < MIN_TIME_TO_RESTART_SLEEP_TIMER_IN_MINUTES && !isSleepEndOfChapterRunning && lastSleepAfterEndOfChapterTime != null
-
-    private fun createAlarm(timeMs: Long): Boolean {
-        val sleepIntent = getSleepIntent()
-        val alarmManager = getAlarmManager()
-        alarmManager.cancel(sleepIntent)
-        return try {
-            LogBuffer.i(TAG, "Starting...")
-            alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, timeMs, sleepIntent)
-            sleepTimeMs = timeMs
-            lastTimeSleepTimeHasFinished = timeMs.milliseconds
-            true
-        } catch (e: Exception) {
-            LogBuffer.e(LogBuffer.TAG_CRASH, e, "Unable to start sleep timer.")
-            false
+        LogBuffer.i(LogBuffer.TAG_PLAYBACK, "Sleeping playback for end of episode")
+
+        onSleepEndOfEpisode()
+    }
+
+    suspend fun sleepEndOfChapter(onSleepEndOfChapter: suspend () -> Unit) {
+        if (getState().isSleepEndOfChapterRunning) {
+            updateSleepTimer {
+                copy(numberOfChaptersLeft = getState().numberOfChaptersLeft - 1)
+            }
+            setEndOfChapter()
         }
+
+        if (getState().isSleepEndOfChapterRunning) return
+
+        updateSleepTimerStatus(sleepTimeRunning = false)
+
+        LogBuffer.i(LogBuffer.TAG_PLAYBACK, "Sleeping playback for end of chapters")
+
+        onSleepEndOfChapter()
     }
 
     fun cancelTimer() {
         LogBuffer.i(TAG, "Cleaning automatic sleep timer feature...")
-        getAlarmManager().cancel(getSleepIntent())
-        cancelSleepTime()
+        updateSleepTimerStatus(sleepTimeRunning = false, sleepAfterChapters = 0, sleepAfterEpisodes = 0)
         cancelAutomaticSleepAfterTimeRestart()
         cancelAutomaticSleepOnEpisodeEndRestart()
         cancelAutomaticSleepOnChapterEndRestart()
     }
 
-    val isSleepAfterTimerRunning: Boolean
-        get() = System.currentTimeMillis() < (sleepTimeMs ?: -1)
-
-    fun timeLeftInSecs(): Int? {
-        val sleepTimeMs = sleepTimeMs ?: return null
-
-        val timeLeft = sleepTimeMs - System.currentTimeMillis()
-        if (timeLeft < 0) {
-            LogBuffer.i(TAG, "Cancelled because time is up")
-            cancelSleepTime()
-            return null
-        }
-        return (timeLeft / DateUtils.SECOND_IN_MILLIS).toInt()
+    private fun setEndOfEpisodeUuid(uuid: String) {
+        LogBuffer.i(TAG, "Episode $uuid was marked as end of episode")
+        sleepTimerHistory = sleepTimerHistory.copy(
+            lastEpisodeUuidAutomaticEnded = uuid,
+            lastTimeSleepTimeHasFinished = System.currentTimeMillis().milliseconds,
+        )
```

Comment:
The change here is that now we have the `sleepTimerHistory`, instead of having separated variables like `lastEpisodeUuidAutomaticEnded` and `lastTimeSleepTimeHasFinished`